#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/RectangleArena.proto"

WorldInfo {
}
Viewpoint {
  orientation 0.5261160967522485 0.014385628872696388 -0.8502910715867336 3.146138113734271
  position 8.666925959588204 3.5046928857722666 11.43106021247259
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Robot {
  translation 2.9276680914435307 -0.15053358241645579 -0.010279444426608995
  rotation -0.001956962460827615 -0.001936082611844881 -0.9999962109338449 1.5695484648147768
  children [
    DEF compass Compass {
    }
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0.917647 0.145098 0.145098
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.2 0.1 0.05
      }
    }
    DEF WHEEL1 HingeJoint {
      jointParameters HingeJointParameters {
        position 218.17924935788312
        axis 0 1 0
        anchor 0.05 0.06 0
      }
      device [
        PositionSensor {
          name "encoder1"
        }
        RotationalMotor {
          name "wheel1"
        }
      ]
      endPoint Solid {
        translation 0.04999776149557196 0.060000001823789756 3.2228990720844696e-06
        rotation 0.24448295179226487 0.6856498344066125 -0.6856474246003743 2.662035291822903
        children [
          DEF WHEEL Shape {
            appearance PBRAppearance {
              baseColor 0.305882 0.898039 0.25098
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.02
              radius 0.04
              subdivision 24
            }
          }
        ]
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL2 HingeJoint {
      jointParameters HingeJointParameters {
        position 314.7720944049624
        axis 0 1 0
        anchor 0.05 -0.06 0
      }
      device [
        PositionSensor {
          name "encoder2"
        }
        RotationalMotor {
          name "wheel2"
        }
      ]
      endPoint Solid {
        translation 0.05000312278835569 -0.059999998174488686 -2.376082334726549e-06
        rotation 0.3309757114054394 0.6672525037928017 -0.6672549547526048 2.502340452955396
        children [
          USE WHEEL
        ]
        name "solid(1)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL3 HingeJoint {
      jointParameters HingeJointParameters {
        position 218.17928364395982
        axis 0 1 0
        anchor -0.05 0.06 0
      }
      device [
        RotationalMotor {
          name "wheel3"
        }
      ]
      endPoint Solid {
        translation -0.05000223851694874 0.05999999817434007 3.222890629009214e-06
        rotation 0.5153065196534552 -0.6059936011142857 0.6059958301929008 2.189960236252736
        children [
          USE WHEEL
        ]
        name "solid(2)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL4 HingeJoint {
      jointParameters HingeJointParameters {
        position 314.7721289627392
        axis 0 1 0
        anchor -0.05 -0.06 0
      }
      device [
        RotationalMotor {
          name "wheel4"
        }
      ]
      endPoint Solid {
        translation -0.04999687720188152 -0.060000001823938484 -2.3760697839563216e-06
        rotation 0.9127894625133761 0.28880337737203565 -0.2888044430805167 1.661923642640378
        children [
          USE WHEEL
        ]
        name "solid(3)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF DS_RIGHT DistanceSensor {
      translation 0.1 -0.03 0
      rotation 0 0 1 -0.3
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.184314 0.596078 0.847059
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_right"
    }
    DEF DS_LEFT DistanceSensor {
      translation 0.1 0.03 0
      rotation 0 0 1 0.3
      children [
        Shape {
          appearance PBRAppearance {
            baseColor 0.184314 0.596078 0.847059
            roughness 1
            metalness 0
          }
          geometry Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_left"
    }
  ]
  boundingObject USE BODY
  physics Physics {
    density -1
    mass 1
  }
  controller "robotcar"
}
Robot {
  translation 2.64297 2.88092 0.136585
  rotation 0 0 1 3.14159
  children [
    HingeJoint {
      jointParameters HingeJointParameters {
        position 592.6950400000034
        axis 0 0 1
        anchor 0 0.00280235 0.070858
      }
      device [
        RotationalMotor {
          name "lidarMotor"
        }
      ]
      endPoint Lidar {
        translation -6.743469518382643e-07 0.002801943476228687 -0.15914200000000017
        rotation 1.732898535921257e-06 4.4693733391776244e-07 0.9999999999983986 3.6464174257477246
        children [
          Compass {
            name "lidarCompass"
          }
          Shape {
            appearance PBRAppearance {
              metalness 0
            }
            geometry Cylinder {
              height 0.03
              radius 0.05
            }
          }
        ]
        horizontalResolution 1
        fieldOfView 0.1
        numberOfLayers 1
        maxRange 4
      }
    }
    DEF compass Compass {
    }
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0 0.145098 0.145098
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.2 0.2 0.3
      }
    }
    DEF WHEEL1 HingeJoint {
      jointParameters HingeJointParameters {
        position 128.5269178416145
        axis 0 1 0
        anchor 0.06 0.103547 -0.149772
      }
      device [
        PositionSensor {
          name "encoder1"
        }
        RotationalMotor {
          name "wheel1"
        }
      ]
      endPoint Solid {
        translation 0.06 0.103547 -0.149772
        rotation 0.7037803004056861 0.5200802341701776 -0.4839523104463659 1.9322548290017059
        children [
          DEF WHEEL Shape {
            appearance PBRAppearance {
              baseColor 0.305882 0.898039 0.25098
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.02
              radius 0.04
              subdivision 24
            }
          }
        ]
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL2 HingeJoint {
      jointParameters HingeJointParameters {
        position 139.70156892824377
        axis 0 1 0
        anchor 0.06 -0.11 -0.149917
      }
      device [
        PositionSensor {
          name "encoder2"
        }
        RotationalMotor {
          name "wheel2"
        }
      ]
      endPoint Solid {
        translation 0.06 -0.11 -0.149917
        rotation 0.022015537112643063 0.706935071536478 -0.7069357260439997 3.0975675402212106
        children [
          USE WHEEL
        ]
        name "solid(1)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL3 HingeJoint {
      jointParameters HingeJointParameters {
        position 128.52696466387883
        axis 0 1 0
        anchor -0.06 0.11 -0.15
      }
      device [
        RotationalMotor {
          name "wheel3"
        }
      ]
      endPoint Solid {
        translation -0.06 0.11 -0.15
        rotation 0.09856399017721049 0.7036623766512711 -0.7036649767651019 2.945099834025829
        children [
          USE WHEEL
        ]
        name "solid(2)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
    DEF WHEEL4 HingeJoint {
      jointParameters HingeJointParameters {
        position 139.70158231605558
        axis 0 1 0
        anchor -0.06 -0.11 -0.15
      }
      device [
        RotationalMotor {
          name "wheel4"
        }
      ]
      endPoint Solid {
        translation -0.06 -0.11 -0.15
        rotation 0.6155450051300749 0.5572720670548054 -0.5572720968610222 2.038048341219826
        children [
          USE WHEEL
        ]
        name "solid(3)"
        boundingObject USE WHEEL
        physics Physics {
        }
      }
    }
  ]
  name "Lidar Scanner"
  boundingObject USE BODY
  physics Physics {
    density -1
    mass 1
  }
  controller "slam"
}
Solid {
  translation 3.3 0.05 0
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall"
  boundingObject USE WALL
}
Solid {
  translation 3.1 0.25 0
  rotation 0 0 1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(1)"
  boundingObject USE WALL
}
Solid {
  translation 2.25 0.5 0
  rotation 0 0 1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(7)"
  boundingObject USE WALL
}
Solid {
  translation 2.25 1 0
  rotation 0 0 1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(10)"
  boundingObject USE WALL
}
Solid {
  translation 2.59 0.75 0
  rotation 0 0 1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(8)"
  boundingObject USE WALL
}
Solid {
  translation 3.3 -0.3 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(2)"
  boundingObject USE WALL
}
Solid {
  translation 2.8 0.55 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(3)"
  boundingObject USE WALL
}
Solid {
  translation 2.8 0.95 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.4
      }
    }
  ]
  name "wall(9)"
  boundingObject USE WALL
}
Solid {
  translation 3.3 0.95 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(13)"
  boundingObject USE WALL
}
Solid {
  translation 2.8 -0.3 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(4)"
  boundingObject USE WALL
}
Solid {
  translation 2.75 -5.0981e-08 0
  rotation 0 0 1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(5)"
  boundingObject USE WALL
}
Solid {
  translation 2.55 0.2 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(6)"
  boundingObject USE WALL
}
Solid {
  translation 2.55 1.3 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(11)"
  boundingObject USE WALL
}
Solid {
  translation 3.05 1.3 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.4
      }
    }
  ]
  name "wall(12)"
  boundingObject USE WALL
}
Solid {
  translation 1.48 2.21 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 3.7 0.1 0.4
      }
    }
  ]
  name "bigwall"
  boundingObject USE WALL
}
Solid {
  translation 0.36 4.66 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 1.3 0.1 0.4
      }
    }
  ]
  name "bigwall(6)"
  boundingObject USE WALL
}
Solid {
  translation 1.38 3.3 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 1.3 0.1 0.4
      }
    }
  ]
  name "bigwall(1)"
  boundingObject USE WALL
}
Solid {
  translation 1.36 6.05 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 4 0.1 0.4
      }
    }
  ]
  name "bigwall(10)"
  boundingObject USE WALL
}
Solid {
  translation 3.92 4.05 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 1.3 0.1 0.4
      }
    }
  ]
  name "bigwall(4)"
  boundingObject USE WALL
}
Solid {
  translation 3.92 2.7 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 1.3 0.1 0.4
      }
    }
  ]
  name "bigwall(8)"
  boundingObject USE WALL
}
Solid {
  translation 3.31 2.41 0
  rotation 0 0 -1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.6 0.1 0.4
      }
    }
  ]
  name "bigwall(2)"
  boundingObject USE WALL
}
Solid {
  translation 3.31 5 0
  rotation 0 0 -1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 2 0.1 0.4
      }
    }
  ]
  name "bigwall(5)"
  boundingObject USE WALL
}
Solid {
  translation -0.32 4.09 0
  rotation 0 0 -1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 3.9 0.1 0.4
      }
    }
  ]
  name "bigwall(3)"
  boundingObject USE WALL
}
Solid {
  translation 2.05 4.52 0
  rotation 0 0 -1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 1.5 0.1 0.4
      }
    }
  ]
  name "bigwall(7)"
  boundingObject USE WALL
}
Solid {
  translation 4.51 3.37 0
  rotation 0 0 -1 -1.5707953071795862
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 1.3 0.1 0.4
      }
    }
  ]
  name "bigwall(9)"
  boundingObject USE WALL
}
Solid {
  translation 3.3 1.3 0
  rotation 0 0 1 3.14159
  children [
    DEF WALL Shape {
      appearance PBRAppearance {
      }
      geometry Box {
        size 0.5 0.1 0.1
      }
    }
  ]
  name "wall(14)"
  boundingObject USE WALL
}
RectangleArena {
  translation 0 0 -0.05
  floorSize 15 15
}
Solid {
  translation 0.13 5.37 -0.06
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 1 0 0
        roughness 1
        IBLStrength 50
      }
      geometry Box {
        size 0.1 0.1 0.1
      }
    }
  ]
  name "destination"
}
